ZXY6005S.py

```diff
@@ -0,0 +1,159 @@
+
+import serial
+import serial.tools.list_ports
+from enum import Enum
+
+class Commands(Enum):
+    '''Get power supply unit model. Response: ZXY-6005S'''
+    MODEL = 'a'
+    '''Get firmware version. Response: R2.7Z'''
+    FIRMWARE_VERSION = 'v'
+    '''Set amp hour counter to specified value. Response: Asa(value)'''
+    SET_AMP_HOUR = 'sa'
+    '''Return the amp hour reading. Response: Ara(5 digit value)'''
+    RETURN_AMP_HOUR = 'ra'
+    '''Set voltage to specified value. Response: Asu(value)'''
+    SET_VOLTAGE = 'su'
+    '''Return voltage measurement. Response: Aru(5 digit value)'''
+    RETURN_VOLTAGE = 'ru'
+    '''Set current limit to specified value. Response: Asi(value)'''
+    SET_CURRENT_LIMIT = 'si'
+    '''Return current in amps. Response: Ari(4 digit value)'''
+    RETURN_CURRENT = 'ri'
+    '''Return Mode [Constant Voltage(CV) or Constant Current(CC)] Response: Arc0 or Arc1'''
+    RETURN_MODE = 'rc'
+    '''Return temperature in Celsius. Response: Art(3 digit value)'''
+    RETURN_TEMP = 'rt'
+    '''Set power output (On/OFF). Response: Aso(1 OR 0)'''
+    SET_OUTPUT = 'so'
+
+
+class ZXY6005S:
+    BAUDRATE = 9600
+    INPUT_DELAY = 0.001
+    BYTESIZE = serial.EIGHTBITS
+    PARITY = serial.PARITY_NONE
+    STOPBITS = serial.STOPBITS_ONE
+    TIMEOUT = 1
+
+    def __init__(self):
+        '''construct objects, using location for X, Y, and Z power supplies'''
+        names = ['X', 'Y', 'Z']
+        locations = ['1-1.5.4.3', '1-1.5.4.2', '1-1.5.4.1']
+        self.devices = {}
+        for name, location in zip(names,locations):
+            serial_port = None
+            for i in serial.tools.list_ports.comports():
+                if i.location == location:
+                    serial_port = i.device
+                    break
+            if serial_port is None:
+                raise Exception(f'Could not find device with location of {location}')
+
+            self.devices[name] = serial.Serial(
+                port = serial_port,
+                baudrate = self.BAUDRATE,
+                parity = self.PARITY,
+                stopbits = self.STOPBITS,
+                bytesize = self.BYTESIZE,
+                timeout = self.TIMEOUT,
+            )
+
+    def write_message(self, device_name, msg):
+        '''writes a command to serial port'''
+        ser = self.devices[device_name]
+        if ser.out_waiting != 0:
+            ser.flush()
+        ser.write(msg)
+        ser.flush()
+
+    def read_message(self, device_name, ending_token = '\n'):
+        '''reads from the serial port until a specified character'''
+        ser = self.devices[device_name]
+        data = ser.read_until(ending_token)
+        return data.decode().strip()
+
+    def send_command(self, device_name, msg):
+        '''sends a command to serial port and reads the message returned'''
+        self.write_message(device_name, msg)
+        return self.read_message(device_name)
+
+    def create_command(self, command):
```

Comment:
This is actually excellent that you've separated the command generation from the command sending.

The core functionality of this class, (i.e. directly communicating with the power supply on a serial channel), is not really testable. However the most important part that we can unit-test is making sure that the commands are being generated correctly per the power-supply spec.

I would recommend making a set of unit tests around this function asserting that this function generates all of the correct things and fails safely in a way that will stop it from being sent to the PSU otherwise.